extensions/amp-gwd-animation/0.1/amp-gwd-animation-impl.js

```diff
@@ -0,0 +1,447 @@
+    const counterName = `${eventName}_${label}`;
+    const currentCount = getOrInitCounter(receiver, counterName);
+
+    if (currentCount < count) {
+      this.playLabelAnimation_(receiver, label);
+      setCounter(receiver, counterName, currentCount + 1);
+    }
+  }
+
+  /**
+   * Returns the element identified by a receiver id if it exists in the
+   * invocation origin document and has a classList. If not found, returns null.
+   * @param {string} id The receiver id.
+   * @return {?Element}
+   */
+  getReceiver(id) {
+    if (id == 'document.body') {
+      return this.ampdoc_.getBody();
+    }
+
+    // Try to locate the receiver by id in the DOM.
+    // TODO(sklobovskaya): When support for groups is added, this lookup will
+    // need to use GwdIds.
+    const receiver = this.ampdoc_.getRootNode().getElementById(id);
+
+    // Check that a valid element was found.
+    if (receiver && receiver.classList) {
+      return receiver;
+    } else {
+      user().warn(LOG_ID, `Could not get receiver with id ${id}.`);
+      return null;
+    }
+  }
+
+  /**
+   * Switches the given element's animation to the given label animation.
+   * This is a core gotoAndPlay routine that is used in all goto* actions.
+   * @param {!Element} receiver
+   * @param {string} label
+   * @private
+   */
+  playLabelAnimation_(receiver, label) {
+    // Unpause playback.
+    receiver.classList.remove(PlaybackCssClass.PAUSE);
+
+    // If another goto animation is currently active on this element, stop it.
+    const currentLabel =
+        receiver.getAttribute(CURRENT_LABEL_ANIMATION_ATTR);
+
+    if (currentLabel) {
+      receiver.classList.remove(currentLabel);
+      receiver.removeAttribute(CURRENT_LABEL_ANIMATION_ATTR);
+    }
+
+    // A forced reflow is needed if removing and readding the same class so
+    // the animation is restarted.
+    if (currentLabel == label) {
+      reflow(receiver);
+    }
+
+    // Add the label animation class and record this label as this element's
+    // current animation.
+    receiver.classList.add(label);
+    receiver.setAttribute(CURRENT_LABEL_ANIMATION_ATTR, label);
+  }
+
+  /**
+   * Handles GWD event `animationend` events, which signal that a timeline event
+   * marker has been reached.
+   * If the event originated from a GWD event element, extracts its event name
+   * and dispatches a custom event.
+   * @param {!Event} event An `animationend` event.
+   * @private
+   */
+  onAnimationEndEvent_(event) {
+    const userEventName = event.target.getAttribute(EVENT_NAME_ATTR);
+
+    if (!userEventName) {
+      // No GWD event name could be extracted, usually just because the event is
+      // from some other animation. Just ignore it.
+      return;
+    }
+
+    const timelineEvent = createCustomEvent(
+        this.ampdoc_.win,
+        `${GWD_TIMELINE_EVENT}`,
+        {eventName: userEventName, sourceEvent: event});
+
+    this.ampdoc_.win.dispatchEvent(timelineEvent);
```

Comment:
This could be tricky. Where will the event go in this case? Will we avoid it being shared to other embeds or documents?

Reply:
Right, this event should just be scoped to the ampdoc root. Done.